x ## 一、建立参考基因组索引

1、下载参考基因组和注释文件

2、构建适用于cellranger分析所用的参考基因组和注释文件索引（index），代码如下：
```bash
cellranger mkgtf chr_susScr11.gtf chr_susScr11.filtered.gtf --attribute=gene_biotype:protein_coding

##根据自己分析的物种参考基因组和注释文件构建gtf索引
cellranger mkref --genome=chr_susScr11_cellranger \
         --nthreads=10 \
         --fasta=chr_susScr11.fa \
         --genes=chr_susScr11.filtered.gtf
```
cellranger mkgtf /bio/hxl/06/06_all_sorted.gtf /bio/hxl/06/cell/06_all_sorted_gai.gtf --attribute=gene_biotype:protein_coding  

cellranger mkref --genome=sc_06macfas --fasta=/bio/hxl/06/cell/06-macFas5.fasta --genes=/bio/hxl/06/cell/06_all_sorted_0gai1.gtf --nthreads=60

cellranger mkref   --genome=06macaque   --nthreads=60   --fasta=/bio/hxl/06/GCA_011100615.1_Macaca_fascicularis_6.0_genomic.fna   --genes=/bio/hxl/06/06_all_sorted.gtf

## 一．将sra文件转化切割为SRR文件_1,_2,_3

fastq-dump -O /data3/hxl/single-cell --split-files /data3/hxl/single-cell/SRR21138448 (不能用就用/zhlab/software/sratoolkit/sratoolkit.3.0.0-centos_linux64/bin/fastq-dump -O /data3/hxl/single_cell-801 --split-files /data3/hxl/single_cell-801/SRR11034803） 批量处理

dump=/zhlab/software/sratoolkit/sratoolkit.3.0.0-centos_linux64/bin/fasterq-dump ls * | while read id; do (nohup $dump -3 -e 30 -p -S -t /zhlab -O ./ /data2/pan-vert/vert-hic/{id} & );done
最终版
/zhlab/software/sratoolkit/sratoolkit.3.0.0-centos_linux64/bin/fastq-dump -O /bio/hxl/06/scrna/ --split-files /bio/hxl/06/scrna/SRR18039374
## 二．将对应的SRR文件名改为!

![screen-capture](file:///E:/markdown%E5%AD%98%E6%A1%A3/%E5%8D%95%E7%BB%86%E8%83%9E%E6%B5%8B%E5%BA%8F%E7%AC%AC%E4%B8%80%E6%AD%A5%E6%B5%81%E7%A8%8B/1cd7dee302f5242499900e9aa9be2f54.png?lastModify=1709708850)

![[Pasted image 20240807091453.png]]
有四个fastq的时候





## 三．使用软件，跑出数据

nohup /bio/hxl/cellranger-8.0.1/bin/cellranger count --id=SRR18042389 --transcriptome=/bio/hxl/06/cell/rhesus06/  --fastqs=/bio/hxl/06/ --sample=SRR18042389 --localcores=60 --localmem=120 --include-introns false --create-bam false --chemistry=ARC-v1 >/bio/hxl/123.log 2>&1 &


/bio/hxl/cellranger-8.0.1/bin/cellranger count --id=SRR18039374 --transcriptome=/bio/hxl/06/sc_06macfas/  --fastqs=/bio/hxl/06/scrna/ --sample=SRR18039374 --localcores=60 --localmem=120 --include-introns false --create-bam true chemistry=auto

如果出现下列问题

Log message:
An extremely low rate of correct barcodes was observed for all the candidate chemistry choices for the input: Sample SRR18042389 in "/bio/hxl/06". Please check your input data.
- 0.2% for chemistry SC3Pv3
- 0.2% for chemistry SC3Pv3HT
- 0.2% for chemistry SC3Pv4
- 0.1% for chemistry SC5P-PE-v3
- 0.0% for chemistry ARC-v1
- 0.0% for chemistry SC5P-PE
- 0.0% for chemistry SC3Pv2
- 0.0% for chemistry SC3Pv3LT
--chemistry <CHEM>
          Assay configuration. NOTE: by default the assay configuration is detected
          automatically, which is the recommended mode. You usually will not need to specify a
          chemistry. Options are: 'auto' for autodetection, 'threeprime' for Single Cell 3',
          'fiveprime' for  Single Cell 5', 'SC3Pv1' or 'SC3Pv2' or 'SC3Pv3' or 'SC3Pv4' for
          Single Cell 3' v1/v2/v3/v4, 'SC3Pv3LT' for Single Cell 3' v3 LT, 'SC3Pv3HT' for
          Single Cell 3' v3 HT, 'SC5P-PE' or 'SC5P-PE-v3' or 'SC5P-R2' or 'SC5P-R2-v3', for
          Single Cell 5', paired-end/R2-only, 'SC-FB' for Single Cell Antibody-only 3' v2 or
          5'. To analyze the GEX portion of multiome data, chemistry must be set to 'ARC-v1'
          [default: auto]


将参数改为--chemistry=ARC-v1






合并aggr流程

需要将library_id改为sample_id

![截图](file:///E:/markdown%E5%AD%98%E6%A1%A3/%E5%8D%95%E7%BB%86%E8%83%9E%E6%B5%8B%E5%BA%8F%E7%AC%AC%E4%B8%80%E6%AD%A5%E6%B5%81%E7%A8%8B/2465b8526c25afc628bd4d4ed1b69874.png?lastModify=1709708850)

/data3/hxl/cellranger-7.0.1/bin/cellranger aggr --id=w1w3 --csv=hebing.csv --normalize=mapped

---

（草稿）

/zhlab/software/sratoolkit/sratoolkit.3.0.0-centos_linux64/bin/fastq-dump -O /data3/hxl/single_cell-801 --split-files /data3/hxl/single_cell-801/SRR11034803

cellranger count --id=cancer_801 --fastqs=/data3/wangshibo --sample=SRR11034801 --transcriptome=/data3/wangshibo/hg38-HERV-cellranger

export PATH=~/download/cellranger-7.0.1:$PATH

 sam-dump SRR21138452.sra | samtools view -bS - > SRR21138452.bam          samtools view SRR21138452.bam | less -SN samtools view SRR21138452.bam | head -3 | tr "\t" "\n" | cat -n

/data3/wangshibo/cellranger-7.0.1/bin/cellranger count  --fastqs=/data3/wangshibo --sample=SRR21138445 --transcriptome==/data3/hg38-HERV-cellranger-2 --id=SRR21138445-count --localcores=60 --localmem=120

ls SRR12164439 |/zhlab/software/sratoolkit/sratoolkit.3.0.0-centos_linux64/bin/fastq-dump --gzip --split-3 -O ./ SRR12164439

SRR11197826_S1_L001_R1_001.fastq.gz

nohup /data3/hxl/cellranger-7.0.1/bin/cellranger count --id=A823-single-cell --transcriptome=/data3/wangshibo/hg38-HERV-cellranger --fastqs=/data3/hxl/823/ --sample=SRR11197823,SRR11197824,SRR11197825,SRR11197826,SRR11197827,SRR11197828,SRR11197829,SRR11197830,SRR11197831,SRR11197832,SRR11197833,SRR11197834 --localcores=60 --localmem=120 --r2-length 91 --include-introns false >/data3/hxl/823_834.log 2>&1 &

SRR11197823_S1_L001_R2_001.fastq.gz SRR11197824_S1_L001_R2_001.fastq.gz SRR11197825_S1_L001_R2_001.fastq.gz SRR11197826_S1_L001_R2_001.fastq.gz SRR11197827_S1_L001_R2_001.fastq.gz SRR11197828_S1_L001_R2_001.fastq.gz SRR11197829_S1_L001_R2_001.fastq.gz SRR11197830_S1_L001_R2_001.fastq.gz SRR11197831_S1_L001_R2_001.fastq.gz SRR11197832_S1_L001_R2_001.fastq.gz SRR11197833_S1_L001_R2_001.fastq.gz SRR11197834_S1_L001_R2_001.fastq.gz

qc命令

ls *.fastq | while read id; do (nohup fastqc ${id} /data3/hxl/fastqc.log 2>&1 &);done

nohup fastqc SRR14327246.lite.1_1.fastq SRR14327246.lite.1_2.fastq SRR14327247.lite.1_1.fastq SRR14327247.lite.1_2.fastq SRR14327248.lite.1_1.fastq SRR14327248.lite.1_2.fastq SRR14327249.lite.1_1.fastq SRR14327249.lite.1_2.fastq SRR14327250.lite.1_1.fastq SRR14327250.lite.1_2.fastq SRR14327251.lite.1_1.fastq SRR14327251.lite.1_2.fastq SRR14327252.lite.1_1.fastq SRR14327252.lite.1_2.fastq SRR14327253.lite.1_1.fastq SRR14327253.lite.1_2.fastq

nohup /data3/hxl/cellranger-7.0.1/bin/cellranger count --id=A831-834-single-cell --transcriptome=/data3/wangshibo/hg38-HERV-cellranger --fastqs=/data3/hxl/823/ --sample=SRR11197831,SRR11197832,SRR11197833,SRR11197834 --localcores=60 --localmem=120 --r2-length 91 --r1-length 28 --include-introns false >/data3/hxl/831-834.log 2>&1 &

nohup /data3/hxl/cellranger-7.0.1/bin/cellranger count --id=A827-830-single-cell --transcriptome=/data3/wangshibo/hg38-HERV-cellranger --fastqs=/data3/hxl/823/ --sample=SRR11197827,SRR11197828,SRR11197829,SRR11197830 --localcores=60 --localmem=120 --r2-length 91 --r1-length 28 --include-introns false >/data3/hxl/827-830.log 2>&1 &




fastq-dump -O /bio/hxl/GSE98422_srr/fa/ --split-files /bio/hxl/GSE98422_srr/SRR5494044